tools/pkg/engine_repo_tools/lib/engine_repo_tools.dart

```diff
@@ -66,25 +72,44 @@ final class Engine {
     // don't want to fail if it doesn't exist.
     final io.Directory outDir = io.Directory(p.join(srcPath, 'out'));
 
-    return Engine._(srcDir, flutterDir, outDir);
+    return Engine._(
+      srcDir: srcDir,
+      flutterDir: flutterDir,
+      outDir: outDir,
+    );
   }
 
   /// Creates an [Engine] by looking for a `src/` directory in the given path.
   ///
+  /// Similar to [tryFindWithin], but throws a [StateError] if the path is not
+  /// within a valid engine. This is useful for tools that require an engine
+  /// and do not have a reasonable fallback or recovery path.
+  factory Engine.findWithin([String? path]) {
```

Comment:
Should `findWithin` and `tryFindWithin` have unit tests?

Reply:
They do:

https://github.com/flutter/engine/blob/ab30b6812f12ddcb342fdf4a88e8bded2ef59f60/tools/pkg/engine_repo_tools/test/engine_repo_tools_test.dart#L80-L159

All that changed is:

- `tryFindWithin` return `Engine?`
- `findWithin` throws if `tryFIndWithin` would return `null`

Otherwise all of the code paths are still tested AFAICT.